org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/ProblemReporter.java

```diff
@@ -12453,4 +12458,40 @@ public void falseLiteralInGuard(Expression exp) {
 			exp.sourceStart,
 			exp.sourceEnd);
 }
+public boolean scheduleProblemForContext(Runnable problemComputation) {
+	if (this.referenceContext != null) {
+		CompilationResult result = this.referenceContext.compilationResult();
+		if (result != null) {
+			try (this) {
+				ReferenceContext capturedContext = this.referenceContext;
+				result.scheduleProblem(() -> {
+					ReferenceContext save = this.referenceContext;
+					this.referenceContext = capturedContext;
+					try {
+						problemComputation.run();
+					} finally {
+						this.referenceContext = save;
+					}
+				});
+			}
+			return true;
+		}
+	}
+	return false;
+}
+/**
+ * General strategy:
+ * <ul>
+ * <li>clients are responsible for assigning {@link #referenceContext} prior to invoking any
+ * reporting method (usually, by obtaining an instance using {@link Scope#problemReporter()}).
+ * <li>individual methods should ensure that {@link #referenceContext} is reset to {@code null} afterwards. This may
+ * happen in two ways:
+ * <ul><li>by calling any of the {@code handle(..)} methods on all paths, or failing that
+ * <li>wrap their body in {@code try(this) { ... }}.
+ * </ul></ul>
+ */
+@Override
+public void close() {
+	this.referenceContext = null;
+}
```

Comment:
@srikanth-sankaran @jarthana One more change that tries to make an implicit strategy more explicit. Do you want to comment on changes outlined in the above javadoc?

I briefly thought of actually pulling all this into a new method in `Scope`, to replace all those `scope.getProblemReporter().reportXYZ(foo,bar)` calls with smth like this:
```
scope.withContext(pr -> pr.reportXYZ(foo,bar));
```
This would encapsulate that strategy even better, but it would require a million of edits all over the place ;-P -- so it may be no more than a theoratical exercise ...

Reply:
I should mention that this change was motivated by the observation that (a) my new method `scheduleProblemForContext()` was implemented incompletely, and (b) nobody noticed, because we accidentally benefit from other methods being sloppy wrt `ProblemReporter#referenceContext`, notably all those calls to `problemReporter().validateRestrictedKeywords()`, which normally never reset the context.

Reply:
> @srikanth-sankaran @jarthana One more change that tries to make an implicit strategy more explicit. Do you want to comment on changes outlined in the above javadoc?

Of course I will accept your comments any time, despite having merged this PR.